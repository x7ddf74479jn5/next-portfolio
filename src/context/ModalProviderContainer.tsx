import * as React from "react";
import type { ModalType, State } from "src/hooks/useModalCore";
import { useModalCore } from "src/hooks/useModalCore";

export const ModalStateContext = React.createContext<State | undefined>(undefined);

type ModalDispatch = {
  openModal: (modalType: ModalType, data?: any) => void;
  closeModal: () => void;
};

export const ModalDispatchContext = React.createContext<ModalDispatch | undefined>(undefined);

type Props = {
  initialState?: Partial<State>;
};

export const ModalProviderContainer: React.FC<Props> = ({ children, initialState }) => {
  const { state, openModal, closeModal } = useModalCore(initialState);
  return (
    <ModalStateContext.Provider value={state}>
      <ModalDispatchContext.Provider value={{ openModal, closeModal }}>{children}</ModalDispatchContext.Provider>
    </ModalStateContext.Provider>
  );
};
