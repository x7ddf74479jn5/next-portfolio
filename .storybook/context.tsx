import React from "react";
import { ModalDispatchContext, ModalStateContext } from "../src/context/ModalProviderContainer";

import { StoryFnReactReturnType } from "@storybook/react/dist/ts3.9/client/preview/types";

const initialState = {
  isModalOpen: false,
  modalType: undefined,
  data: null,
};

const mockDispatchContextValue = {
  openModal: () => {
    // eslint-disable-next-line no-console
    console.log("Modal is open");
  },
  closeModal: () => {},
};

export const withContext = (storyFn: () => StoryFnReactReturnType) => {
  return (
    <ModalStateContext.Provider value={initialState}>
      <ModalDispatchContext.Provider value={mockDispatchContextValue}>{storyFn()}</ModalDispatchContext.Provider>
    </ModalStateContext.Provider>
  );
};
