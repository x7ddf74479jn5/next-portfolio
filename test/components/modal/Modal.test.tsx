import type Queries from "@testing-library/dom/types/queries";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ModalPortal from "src/components/modal/Modal";
import { ModalDispatchContext, ModalStateContext } from "src/context/ModalProviderContainer";
import type { State } from "src/hooks/useModalCore";

const initialState: State = {
  isModalOpen: false,
  modalType: undefined,
  data: null,
};

const mockState = (state?: State): State => {
  return { ...initialState, ...state };
};

const mockDispatchContextValue = {
  openModal: jest.fn(),
  closeModal: jest.fn(),
};

const customRender = (ui: React.ReactElement, state?: State): RenderResult<typeof Queries, HTMLElement> => {
  const mockStateContextValue = mockState(state);
  return render(
    <>
      <div id="__next" />
      <ModalStateContext.Provider value={mockStateContextValue}>
        <ModalDispatchContext.Provider value={mockDispatchContextValue}>{ui}</ModalDispatchContext.Provider>
      </ModalStateContext.Provider>
    </>
  );
};

describe("ModalPortal", () => {
  let renderResult: RenderResult;
  let newState: State;

  it("initial render matches snapshot", () => {
    renderResult = customRender(<ModalPortal />);
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("drawer modal matches snapshot", () => {
    newState = {
      isModalOpen: true,
      modalType: "DRAWER",
      data: null,
    };
    renderResult = customRender(<ModalPortal />, newState);
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("contact modal matches snapshot", () => {
    newState = {
      isModalOpen: true,
      modalType: "CONTACT",
      data: {
        formData: {
          name: "name",
          email: "email@test.com",
          category: "",
          description: "description",
        },
      },
    };
    renderResult = customRender(<ModalPortal />, newState);
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("closes modal when backdrop is clicked", () => {
    newState = {
      isModalOpen: true,
      modalType: "DRAWER",
      data: null,
    };
    renderResult = customRender(<ModalPortal />, newState);

    const backdrop = renderResult.container.getElementsByClassName("backdrop_70")[0];
    expect(backdrop).toBeInTheDocument();
    expect(renderResult.container.childElementCount).toBeGreaterThan(0);

    userEvent.click(backdrop);
    renderResult.rerender(<ModalPortal />);

    expect(renderResult.container.childElementCount).toBe(0);
  });
});
