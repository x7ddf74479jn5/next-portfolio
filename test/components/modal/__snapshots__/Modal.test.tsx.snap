// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalPortal contact modal matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="__next"
  >
    <div
      class="modalContainer"
      role="dialog"
    >
      <div
        class="backdrop_70"
      />
      <div
        class="c-modal-popup"
        id="popup"
      >
        <div
          class="modalPopupBody"
        >
          <h2
            class="modalPopupTitle"
            id="modal-title"
          >
            入力内容の確認
          </h2>
          <div
            class="modalPopupMain"
            id="popup-main"
          >
            <div
              id="mail-content"
            >
              <p>
                お名前：
              </p>
              <p>
                name
              </p>
              <p>
                メールアドレス：
              </p>
              <p>
                email@test.com
              </p>
              <p>
                お問い合わせ種別：
              </p>
              <p />
              <p>
                お問い合わせ内容：
              </p>
              <p>
                description
              </p>
            </div>
          </div>
          <div
            class="gridRow"
          >
            <button
              class="btnCancel"
              id="cancel-btn"
              type="button"
            >
              修正する
            </button>
            <button
              class="btnApply"
              id="apply-btn"
              type="button"
            >
              送信する
            </button>
          </div>
        </div>
      </div>
      <button
        class="closeButton"
      >
        ×
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModalPortal drawer modal matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="__next"
  >
    <div
      class="modalContainer"
      role="dialog"
    >
      <div
        class="backdrop_70"
      />
      <div
        class="modalContent"
      >
        <button>
          close
        </button>
      </div>
      <button
        class="closeButton"
      >
        ×
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModalPortal initial render matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="__next"
  />
</DocumentFragment>
`;
